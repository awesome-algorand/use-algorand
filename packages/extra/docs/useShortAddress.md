# `useShortAddress`

React hook that shortens a wallet address string.

## Usage

```jsx
import { useShortAddress } from 'awesomealgo-hooks';

const Demo = () => {
  const shortAddress = useShortAddress('NRLA7VZ2YV6WOS2LBS3UK25DB463XRSNG63BYNQI6CSGGROFWVQ2EKBQSI');

  return (
    <div>
      shortened address: {shortAddress}
    </div>
  );
};
```

#### Call signature

```typescript
function useShortAddress = (address: string, width: number = 6): string ;
```
