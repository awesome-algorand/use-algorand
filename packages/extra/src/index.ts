export { default as useNfdLookup } from './nfd/useLookup';
export { default as useSubtopiaSubLookup } from './subtopia/useSubscriptionLookup';
export { default as useShortAddress } from './useShortAddress';
export { default as useExplorerUrl } from './useExplorerUrl';
