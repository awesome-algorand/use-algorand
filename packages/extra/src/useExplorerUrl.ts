import { useState, useEffect } from 'react';
import { ExplorerUrlType, ChainType, ExplorerType } from './misc/enums';
import { ALGOEXPLORER_URL, ALGOSCAN_URL } from './misc/constants';

const explorerDefaultUrlPathMapping = (type: string) => {
  switch (type) {
    case ExplorerUrlType.Transaction:
      return `tx`;
    case ExplorerUrlType.Address:
      return `address`;
    case ExplorerUrlType.Asset:
      return `asset`;
    case ExplorerUrlType.Application:
      return `application`;
    default:
      return ``;
  }
};

const explorerBaseUrlMapping = (type: ExplorerType, chain: ChainType) => {
  switch (type) {
    case ExplorerType.AlgoExplorer:
      return ALGOEXPLORER_URL(chain);
    case ExplorerType.AlgoScan:
      return ALGOSCAN_URL(chain);
    default:
      return ``;
  }
};

function useExplorerUrl(
  chain: ChainType,
  input: string | number,
  type: ExplorerUrlType,
  explorerType: ExplorerType = ExplorerType.AlgoScan,
) {
  const [url, setUrl] = useState('');

  useEffect(() => {
    const assembledUrl = `${explorerBaseUrlMapping(
      explorerType,
      chain,
    )}/${explorerDefaultUrlPathMapping(type)}/${input}`;
    setUrl(assembledUrl);
  }, [chain, input, type, explorerType]);

  return url;
}

export default useExplorerUrl;
