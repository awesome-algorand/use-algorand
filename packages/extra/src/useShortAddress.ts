const useShortAddress = (address: string, width: number = 6): string => {
  return `${address.slice(0, width)}...${address.slice(-width)}`;
};

export default useShortAddress;
