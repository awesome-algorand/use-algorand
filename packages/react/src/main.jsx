import React from 'react'
import ReactDOM from 'react-dom/client'
import {QueryClient, QueryClientProvider} from "@tanstack/react-query";
import {ReactQueryDevtools} from "@tanstack/react-query-devtools";
import HomePage from "./routes/HomePage.jsx";
import MethodPage from "./routes/MethodPage.jsx";
import {BrowserRouter, Route, Routes} from "react-router-dom";

const queryClient = new QueryClient()

ReactDOM.createRoot(document.getElementById('root')).render(
    <React.StrictMode>
        <QueryClientProvider client={queryClient}>
            <BrowserRouter>
                <Routes>
                    <Route path="/" element={<HomePage />} />
                    <Route path="/:interface/:method" element={<MethodPage />} />
                </Routes>
            </BrowserRouter>
            <ReactQueryDevtools initialIsOpen={false} />
        </QueryClientProvider>
    </React.StrictMode>,
)
