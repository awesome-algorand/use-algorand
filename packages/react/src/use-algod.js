import {getBlockQueryOptions} from "@algofam/use-algorand-queries";
import { useQuery } from "@tanstack/react-query";

/**
 *
 * @param {number} round
 */
export function useBlock(round){
    return useQuery(getBlockQueryOptions(round))
}
