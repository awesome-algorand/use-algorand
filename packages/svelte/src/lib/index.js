export * from './use-algod.js'
