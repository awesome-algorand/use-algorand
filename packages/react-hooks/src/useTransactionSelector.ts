import {useEffect, useState} from "react";
import { useWallet } from "@txnlab/use-wallet-react";
import {useBlockInformation} from "@awesome-algorand/react-query/lib/algosdk";

/**
 * Transaction Selector Hook
 *
 * Watch the blockchain for a transaction that matches the selector,
 * then return the round that the transaction was found in.
 *
 * @param {Function} selector Filter function
 * @TODO Migrate to algokit-subscriber
 */
export function useTransactionSelector(
    selector: (txns: any[]) => boolean = () => true,
) {
    /**
     * The algorand client
     */
    const manager = useWallet()
    /**
     * The last round that the selector returned true
     */
    const [lastRound, setLastRound] = useState(BigInt(0));
    /**
     * The current round we are checking
     */
    const [round, setRound] = useState(BigInt(0));
    /**
     * The block for the current round
     */
    const block = useBlockInformation({round: Number(round)});

    // Check if the selector returns true for the current block
    useEffect(() => {
        if(typeof block.data === "undefined") return
        let txns = (block.data as any).block.txns
        if (Array.isArray(txns) && txns.length > 0) {
            if (selector(txns)) {
                setLastRound(BigInt(round));
            }
        }
    }, [round, block.data, selector]);

    // Increment the round after next block
    useEffect(() => {
        if (!manager.algodClient) return;
        manager.algodClient
            .statusAfterBlock(round)
            .do()
            .then((b) => setRound(round === BigInt(0) ? b.lastRound : round + BigInt(1)));
    }, [manager.algodClient, round]);

    // Return the last round that the selector returned true
    return lastRound;
}
