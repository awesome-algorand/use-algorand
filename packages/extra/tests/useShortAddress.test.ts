import * as useShortAddress from '../src/useShortAddress';
// @ponicode
describe('useShortAddress.default', () => {
  test('0', () => {
    let result: any = useShortAddress.default('192.168.1.5', 100);
    expect(result).toMatchSnapshot();
  });

  test('1', () => {
    let result: any = useShortAddress.default('192.168.1.5', 64);
    expect(result).toMatchSnapshot();
  });

  test('2', () => {
    let result: any = useShortAddress.default('192.168.1.5', 24);
    expect(result).toMatchSnapshot();
  });

  test('3', () => {
    let result: any = useShortAddress.default('0.0.0.0', 10);
    expect(result).toMatchSnapshot();
  });

  test('4', () => {
    let result: any = useShortAddress.default('192.168.1.5', 16);
    expect(result).toMatchSnapshot();
  });

  test('5', () => {
    let result: any = useShortAddress.default('', -Infinity);
    expect(result).toMatchSnapshot();
  });
});
