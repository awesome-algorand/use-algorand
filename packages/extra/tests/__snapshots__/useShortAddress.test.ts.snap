// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useShortAddress.default 0 1`] = `"192.168.1.5...192.168.1.5"`;

exports[`useShortAddress.default 1 1`] = `"192.168.1.5...192.168.1.5"`;

exports[`useShortAddress.default 2 1`] = `"192.168.1.5...192.168.1.5"`;

exports[`useShortAddress.default 3 1`] = `"0.0.0.0...0.0.0.0"`;

exports[`useShortAddress.default 4 1`] = `"192.168.1.5...192.168.1.5"`;

exports[`useShortAddress.default 5 1`] = `"..."`;
